yt_dlp/extractor/lastfm.py

```diff
@@ -0,0 +1,153 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+import re
+
+from .common import InfoExtractor
+from ..utils import int_or_none
+
+
+class LastFMPlaylistBaseIE(InfoExtractor):
+    def _extract_current_page(self, url):
+        return int_or_none(self._search_regex(
+            r'\bpage=(\d+)', url, 'page', default=None))
+
+    def _extract_last_page(self, webpage):
+        last_page = None
+        pagination_list = self._search_regex(
+            r'(?s)class="[^"]*pagination-list[^"]*"[^>]*>(.+?)</ul>', webpage, 'pagination_list', fatal=False,
+            default=None)
+
+        if pagination_list:
+            last_page = max(map(int_or_none, re.findall(r'(\d+)', pagination_list)))
+
+        return last_page
+
+    def _extract_entries(self, webpage):
+        tbody = self._search_regex(
+            r'(?s)tbody\s+data-playlisting-add-entries[^>]*>(.+?)</tbody>', webpage, 'tbody')
+        return [
+            self.url_result(player_url, 'Youtube')
+            for player_url in re.findall(r'href="([^"]+youtube.com/watch[^"]+)"', tbody)
+        ]
+
+    def _download_page(self, url, current_page_number, last_page_number, playlist_id):
+        note = 'Downloading page'
+        query = {}
+
+        if last_page_number:
+            note = '%s %d of %d' % (note, current_page_number, last_page_number)
+            query['page'] = current_page_number
+
+        return self._download_webpage(url, playlist_id, note, query=query)
+
+    def _entries(self, url, playlist_id):
+        webpage = self._download_webpage(url, playlist_id)
+        last_page_number = self._extract_last_page(webpage)
+        start_page_number = self._extract_current_page(url) or 1
+
+        for page_number in range(start_page_number, (last_page_number or start_page_number) + 1):
+            webpage = self._download_page(url, page_number, last_page_number, playlist_id)
+            page_entries = self._extract_entries(webpage)
+
+            for e in page_entries:
+                yield e
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        return self.playlist_result(self._entries(url, playlist_id), playlist_id)
```

Comment:
urls with 3 directories can be handled by the playlist class, single track and user playlist have 4 directories hence the multi-class approach to avoid duplicate extractor matching issue